pkg/trimaran/resourcestats.go

```diff
@@ -136,3 +154,150 @@ func GetResourceRequested(pod *v1.Pod) *framework.Resource {
 	}
 	return result
 }
+
+// GetResourceLimits : calculate the resource limits of a pod (CPU and Memory); zero limits unchanged, but noted as unLimited
+func GetResourceLimits(pod *v1.Pod) (result *framework.Resource, unlimitedCPU bool, unlimitedMemory bool) {
+	// add up limit of all containers
+	result = &framework.Resource{}
+	for _, container := range pod.Spec.Containers {
+		result.Add(container.Resources.Limits)
+		if container.Resources.Limits.Cpu().IsZero() {
+			unlimitedCPU = true
+		}
+		if container.Resources.Limits.Memory().IsZero() {
+			unlimitedMemory = true
+		}
```

Comment:
Let me clarify.

1. The idea of using `unlimitedCPU` is to capture the fact that if a container specifies CPU requests but no CPU limits, it can potentially surge its usage to 100% node utilization. Of course this depends on what the other containers on the node are doing. The Kubelet will allow it to happen. Right? So, we would rather schedule a new pod on a different node, with limited CPU, than this unlimited node. This way, there would be less risk for the new pod if it needs to burst beyond its requested value, and, at the same time, let this unlimited container burst as well! If scheduled to the same node, then both pods will share the excess capacity (beyond requests) proportional to their requested values. However, if scheduled on different nodes, there is no competition.
2. If an init container does not specify a limit, then again it could burst and use up the full node, if allowed. So, the pod with this init container, as a whole, could use 100% of the node, even only during initialization, which may last a significant amount of time, depending on the appllication. This situation poses a risk to other pods being scheduled to this node. I think this is similar to the SpringBoot scenario?

Reply:
I think the principle for best-efforts Pods is: try your best to tune the limits. What we can try to achieve is that based on the semantics of requests/limits, try to pack more workloads in the cluster while with less disruption/eviction.

In other words, it's impractical to promise users their apps can achieve optimal scheduling behavior by specifying 0 limits. As I mentioned in another comment, giving Pods with 0 limits a "preference" is hazard as users can make use of it.

Reply:
Yes, agreed.